<!--
  Simulated version of Thorp's cat hunter app:
  - Thorp Gazebo simulation
  - Object recognition
  - IPA coverage planning

Explore the entire environment searching for cats, and attack them whenever found.
 -->

<launch>
  <arg name="app_name" value="cat_hunter"/>
  <arg name="viz_smach" default="false"/>
  <arg name="start_delay" default="0.0"/>
  <arg name="record_video" default="false"/>

  <!--  ***** Simulated mobile manipulator Thorp ******  -->
  <include file="$(find thorp_simulation)/launch/thorp_gazebo.launch">
    <arg name="world_name"     value="fun_house"/>
    <arg name="initial_pose_x" value="10.5"/>
    <arg name="initial_pose_y" value="6.0"/>
    <arg name="initial_pose_a" value="0.75"/>
    <arg name="localization"   value="gazebo"/>
  </include>

  <!-- Spawn some cute cats randomly around the house -->
  <node pkg="thorp_simulation" type="spawn_gazebo_models.py" name="objects_spawner" output="screen" args="cats"/>


  <!--  ***** Object recognition using the Kinect *****  -->

  <include file="$(find thorp_perception)/launch/perception.launch">
    <arg name="pipeline" value="COB"/>
    <arg name="target_objects" value="[cat, dog, horse]"/>
  </include>


  <!--  ****************** Utilities ******************  -->

  <node name="model_markers" pkg="thorp_simulation" type="model_markers.py" output="screen" respawn="true"
        args="'cat_orange_0 cat_orange_1 cat_black_0 cat_black_1'"/>

  <node name="cats_controller" pkg="thorp_simulation" type="cats_controller.py" output="screen" respawn="true"
        args="'cat_orange_0 cat_orange_1 cat_black_0 cat_black_1'"/>


  <!--  ****** IPA coverage planning using kinect *****  -->
  <include file="$(find thorp_bringup)/launch/includes/exploration.launch.xml">
    <arg name="camera" value="kinect"/>
  </include>


  <!--  ******* Components common for all apps ********  -->

  <include file="$(find thorp_simulation)/launch/includes/apps_common.launch.xml">
    <arg name="app_name" value="$(arg app_name)"/>
    <arg name="viz_smach" value="$(arg viz_smach)"/>
    <arg name="start_delay" value="$(arg start_delay)"/>
    <arg name="record_video" value="$(arg record_video)"/>
  </include>

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find thorp_bringup)/rviz/hunting.rviz" output="screen" respawn="true"/>
</launch>
